### PR TITLE
Add register pool

### PR DESCRIPTION
**Pull request by guillep**
Introduce the concept of register pools.
A fixed register pool has a limited number of registers.
When no more registers are available, the register allocator will spill the register into a stack slot.

An infinite register pool always has new registers.
However, it will reuse them when possible (not conflicting live ranges).
This leads to an almost optimal (minimal?) usage of registers, depending on the register allocation algorithm of course.

First part of #55

### CODE REVIEW

File: Druid/DRCogitStackToRegisterMappingGenerator.class.st [hunk 1]
```diff
@@ -68,6 +68,13 @@ DRCogitStackToRegisterMappingGenerator >> returnNodeForCompiledMethod [
 	^ RBLiteralNode value: 0
 ]
 
+{ #category : #visiting }
+DRCogitStackToRegisterMappingGenerator >> rtlExpressionForValue: aValue [
+
+	variables at: aValue result ifPresent: [ :var | ^ RBVariableNode named: var ].
+	^ super rtlExpressionForValue: aValue
+]
+
```
Comment (PalumboN) [inline Druid/DRCogitStackToRegisterMappingGenerator.class.st:77]: Why sometimes the order of the methods change? Could be a _non-deterministic situation_ in.... Fuel/Iceberg?

File: Druid/DRInfiniteRegisterPool.class.st [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	currentRegisters := aCollection copy.
+	inactiveOverlappingRegisters do: [ :e |
+		currentRegisters
+			remove: e
+			ifAbsent: [ "not care, we just want to ignore that register..." ] ].
+
+	"If we have registers, take one"
+	currentRegisters ifNotEmpty: [
+		presentBlock value: currentRegisters first ]
+]
```
Comment (PalumboN) [inline Druid/DRInfiniteRegisterPool.class.st:94]: This method is the same for Fixed and Infinite pools.
There are also parts of logic in common between the callers.

Maybe is worth pushing up to a superclass.... maybe not :)

File: Druid/DRLinearScanRegisterAllocator.class.st [hunk 3]
```diff
@@ -4,47 +4,40 @@ Class {
...
+		ifTrue: [
+			registerPool
+				takeFloatingPointRegisterNotIn: inactiveOverlappingRegisters
+				ifPresent: [ :selectedRegister |
+				^ presentBlock value: selectedRegister ] ]
+		ifFalse: [
+			registerPool
+				takeIntegerRegisterNotIn: inactiveOverlappingRegisters
+				ifPresent: [ :selectedRegister |
+				^ presentBlock value: selectedRegister ] ]
```
Comment (PalumboN) [inline Druid/DRLinearScanRegisterAllocator.class.st:40]: Nice! I like more this version 👍

**Review by PalumboN — APPROVED** [review]
Piolaaaaaa 🚀 

I left one comment that you can take for the next iteration

File: Druid/DRCogitStackToRegisterMappingGenerator.class.st (hunk 1 above)
Comment (guillep) [reply Druid/DRCogitStackToRegisterMappingGenerator.class.st:77]: tonel

File: Druid/DRInfiniteRegisterPool.class.st (hunk 2 above)
Comment (guillep) [reply Druid/DRInfiniteRegisterPool.class.st:94]: yes, I prototyped it and I forgot

File: Druid/DRLinearScanRegisterAllocator.class.st (hunk 3 above)
Comment (guillep) [reply Druid/DRLinearScanRegisterAllocator.class.st:40]: It can still be improved but yes

**Pull request merged**